### ISSUE TITLE
Build OpenAPI document from ASP.NET Core's new endpoint routing API

### ISSUE THREAD

**Issue by dustinmoris** [domaindrivendev/Swashbuckle.AspNetCore#1306]
ASP.NET Core has a new endpoint routing API which is being used by MVC and other middleware to register routes in ASP.NET Core. Previously every middleware which needed to do some routing decisions was responsible for implementing it themselves. For example, MVC implemented it with Controllers and Action methods and an attribute based API. An authentication middleware on the otherhand wouldn't know about MVC at all and just implement one or two individual endpoints directly as a `RequestDelegate`. Other frameworks or middlewares had yet their own way of doing routing.

With ASP.NET Core 3 (and going forward) all of these things can/will use the new endpoint routing API, which lets them register a delegate for a given endpoint. Other middleware in the pipeline can retrieve and inspect these routes and make more informed decisons.

It is clear that endpoint routing is the future way of routing in ASP.NET Core and MVC already integrates with it when calling the `MapMvcControllers` extension method during Startup (which will become the default in the future).

I think Swashbuckle.AspNetCore should equally use the endpoint routing API to generate the Swagger/OpenAPI docs (and perhaps slowly move away from the attribute based API, which has a very limited usability, namely MVC only).

Are there any plans or ETA for when this might happen?

**Comment by domaindrivendev** [comment]
5.0.0 is about getting a stable .NET Core 3.0 version out. This doesn't feel critical to that and therefore pushing out. Will try get this done as a fast follow 5.1.0.

**Comment by superstator** [comment]
What is the workaround in the interim? In the past I had a Configure method like this:

```c#
app.UsePathBase("/api/v1");
app.UseMvc();
app.UseSwagger(c =>
{
    c.PreSerializeFilters.Add((swaggerDoc, httpReq) => swaggerDoc.BasePath = "/api/v1");
});
app.UseSwaggerUI(c =>
{
    c.SwaggerEndpoint("/swagger/v1/swagger.json", $"My Cool API");
});
```

and now I have:

```c#
app.UseRouting();
app.UsePathBase("/api/v1");
app.UseSwagger();
app.UseSwaggerUI(c =>
{
    c.SwaggerEndpoint("/swagger/v1/swagger.json", $"My Cool API");
});
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
});
```

But the `/swagger/v1/swagger.json` endpoint returns a 404.

**Comment by akrock** [comment]
@domaindrivendev  Any word on support for endpoint routing? Is this something you are still looking at adding into some 5.x version?

We are looking to adopt endpoint routing in our systems over `UseMvc` now that it's available =)

**Comment by domaindrivendev** [comment]
Although it's not yet documented, this functionality was added by #1619 and is available in the `5.4.0` release

**Comment by stevenknox** [comment]
Does 5.4.0 add support for any endpoint route or specifically just for serving the Swagger JSON endpoint? I had a look at the pull request adding that change but not sure how I can take that and implement for my own routes?

**Comment by jkone27** [comment]
Any examples with F# Giraffe on how to do this? Or with C# endpoints (also ok 💯)

**Comment by dustinmoris** [comment]
@jkone27 We just released `5.0.0-alpha-001` of Giraffe which has a preview of Endpoint routing in an early stage which we can use now to test with the latest version of Swashbuckle. I'll try to put something together over next next coming days!

**Comment by dustinmoris** [comment]
> Although it's not yet documented, this functionality was added (at least for the Swagger JSON endpoint) by #1619 and is available in the 5.4.0 release

@domaindrivendev many thanks!!!

**Comment by panesofglass** [comment]
I've been trying to add this, but I'm running into some strange DI errors:

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HM05I6LSP66J", Request id "0HM05I6LSP66J:00000005": An unhandled exception was thrown by the application. 
System.InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.Mvc.ApiExplorer.IApiDescriptionGroupCollectionProvider' while attempting to activate 'Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator'.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(Type serviceType, Type implementationType, CallSiteChain callSiteChain, ParameterInfo[] parameters, Boolean throwIfCallSiteNotFound)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateConstructorCallSite(ResultCache lifetime, Type serviceType, Type implementationType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, Int32 slot)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateCallSite(Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.<>c__DisplayClass7_0.<GetCallSite>b__0(Type type)  
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.GetCallSite(Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.CreateServiceAccessor(Type serviceType)      
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.GetService(IServiceProvider sp, Type type, Type middleware)
   at lambda_method(Closure , Object , HttpContext , IServiceProvider )
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass4_1.<UseMiddleware>b__2(HttpContext context)        
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

**Comment by slang25** [comment]
There seems to have been a misunderstanding, the endpoint routing support requested in this issue is for the ability to document routing endpoints and have them be picked up by swashbuckle, rather than the ability to expose swagger via endpoint routing (which is still cool btw).

@domaindrivendev with that in mind, is this something that is planned?

**Comment by domaindrivendev** [comment]
@slang25 thanks for pointing out the misunderstanding here.

If the endpoint routing API is going to become the defacto approach for building APIs going forward, then certainly Swashbuckle should support it.

However, Swashbuckle depends heavily on `ApiExplorer`, the metadata component that's built into ASP.NET Core. This component surfaces an `ApiDescription` for every API action method exposed in an application, and Swashbuckle in turn uses this to generate corresponding Swagger descriptions.

So, really this is an upstream issue - specifically for `ApiExplorer` to include operations defined through Endpoint Routing in addition to those defined via MVC/attribute routing.

@rynowak - interested to know if you have any insight on this?

**Comment by rynowak** [comment]
This has been brought up with me a few times - its a "refactor the world" kind of issue. 

@JamesNK has actually done an implementation of this for gRPC - where gRPC endpoints (not controllers) show up as ApiDescriptions. The mechanism is by implementing a custom `IApiDescriptionProvider`. This is a pretty tactical solution since it still references MVC - however given that it's all shared framework at this point does it matter?

Another tactical solution would be for Swashbuckle to add its own extensibility that's suitable for use with endpoints. 

A few thoughts:

Endpoints (and metadata) are computed at startup, and ideally everything that goes into them is **constructable** objects (not services). This is a verbose way of saying that endpoints are **data**. This is relevant because computing the `ApiDescription`(s) for something requires running a lot of expensive code and also customization via services. It's not a great idea to *just* add `ApiDescription` as a metadata, it's probably a limiting choice. You'd probably want some kind of lookup service. 

`ApiDescription` as a type is tied to MVC and that won't really be changeable. In general I am supporting of "pushing down" concepts like `ApiDescription` where possible and making them more general. This is the whole idea behind endpoints to begin with ... So this would look like producing a new exchange type in asp.net core, and a new service for querying them, and having MVC implement that.

Swashbuckle + MVC are a good fit because MVC can gather a lot of swagger-relevant information just from the code you write without extra annotations due to the use of methods for MVC's programming model (vs lambdas). The programming model we surface for using route-to-code directly isn't a good fit for swashbuckle because it doesn't have binding based on parameters nor serialization based on return values. Basically the synergy is lost when you use an imperative rather than declarative programming model. I don't think there's much value in us trying to get good swashbuckle support for route-to-code - it's a better fit for manually writing a swagger doc, or generating stubs from a swagger doc. 

The other major use case is for frameworks like Giraffe, which can't get a swagger story today without plugging into MVC. I think what would be needed here is for someone to come up with a concrete proposal and try to get agreement. 

The major decision seems to be whether the exchange types/contracts go into ASP.NET Core, or swashbuckle/some-other-library (I think the pros/cons are the same if it's outside of ASP.NET Core).

- inside ASP.NET Core: more investment needed but also more reach, MVC can use it
- outside ASP.NET Core: quicker to build, MVC will continue to use ApiDescription

I don't claim to speak for all stakeholders here, so if got something from your POV please speak up 😆 👍

**Comment by rynowak** [comment]
@pranavkm @bradygaster for 👀

**Comment by slang25** [comment]
Thanks for taking the time for a detailed response @rynowak 😄  The `IApiDescriptionProvider` is interesting to me, I wasn't aware of it, and opens up an option for us to potentially integrate with Swashbuckle today, I'll have a play about with that.

In the more medium term, I don't know what others peoples views are, but it seems like pushing down `ApiDescription` for the wider aspnetcore eco-system would be worthwhile. Carter (and Giraffe) for example could make use of it, today Carter builds its OpenAPI docs, which has its pros and cons, but it seems like a sensible thing to standardise across various frameworks/libraries.

**Comment by panesofglass** [comment]
I’m going back to my original idea to use the OpenAPI libraries and, as @rynowak notes, treat it all as data and let the user specify it explicitly, doing some type reflection for JSON Schema generation.

**Comment by JamesNK** [comment]
I integrated gRPC HTTP API endpoints with Swashbuckle using the endpoint routing API here - https://github.com/aspnet/AspLabs/tree/master/src/GrpcHttpApi/src/Microsoft.AspNetCore.Grpc.Swagger

**Comment by dustinmoris** [comment]
@rynowak Thanks for the detailed response with your thoughts and ideas on this topic.

> The major decision seems to be whether the exchange types/contracts go into ASP.NET Core, or swashbuckle/some-other-library

Feels like swashbuckle to me unless these exchange types are going to be used by something else in ASP.NET Core as well, which I don't think there's a use case for it yet.

I think it would be nice if a library like Swashbuckle would have it's own `ApiDoc` kind of type which sits between `ApiDescriptor` and the actual generation of a `swagger.json`:

MVC with its tightly coupled `ApiDescriptor` then gets translated into `ApiDoc` and Swashbuckle produces a Swagger JSON document from `ApiDoc`.

Other non MVC web libraries can then plug/produce into `ApiDoc` without having to use MVC for `ApiDescriptor` and are able to generate an `ApiDoc` kind of contract in an imperative programming model.

**Comment by akrock** [comment]
@domaindrivendev I finally got back around to troubleshooting our implementation. Swashbuckle was behaving as expected and including controllers when using endpoint routing with `MapControllers`. I was able to confirm this in a simplified test project.

Due to the upgrade to .net core 3.1 we also needed to update the `Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer` package from `3.0.0` => `4.1.1`. This upgrade caused issues because of a change in the default behavior introduced in `3.1.0` of this package.

The behavior we were seeing was swagger reporting that no operations were defined and only a single `V1` version was available.

The fix was to revert the change in default behavior introduced in `Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer` by setting `UseApiBehavior` to false in the `ApiVersioningOptions`. With this default value of true, only Controllers that are marked with the `[ApiController]` attribute would be included (which is none of ours).
